Error while checking build/O2/o2-dev-fairroot for e400fe7283b021af3e4996796e51346e481c204c:
```
sw/BUILD/O2-latest/log
[ 30%] Built target test_DataGenerator
Scanning dependencies of target MCStepLogger
[ 31%] Building CXX object Utilities/MCStepLogger/CMakeFiles/MCStepLogger.dir/src/MCStepInterceptor.cxx.o
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Utilities/MCStepLogger/src/MCStepInterceptor.cxx:62:17: error: conflicting declaration of C function 'void dispatchOriginal(TVirtualMagField*, const char*, const char*, const double*, double*)'
 extern "C" void dispatchOriginal(TVirtualMagField*, char const* libname, char const*, const double x[3], double* B);
                 ^~~~~~~~~~~~~~~~
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Utilities/MCStepLogger/src/MCStepInterceptor.cxx:61:17: note: previous declaration 'void dispatchOriginal(TVirtualMCApplication*, const char*, const char*)'
 extern "C" void dispatchOriginal(TVirtualMCApplication*, char const* libname, char const*);
                 ^~~~~~~~~~~~~~~~
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Utilities/MCStepLogger/src/MCStepInterceptor.cxx: In member function 'void o2::field::MagneticField::GetFieldValue(const double*, double*)':
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Utilities/MCStepLogger/src/MCStepInterceptor.cxx:93:62: error: 'dispatchOriginalField' was not declared in this scope
     dispatchOriginalField(baseptr, LIB, SYMBOL, point, bField);    \
                                                              ^
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Utilities/MCStepLogger/src/MCStepInterceptor.cxx:100:1: note: in expansion of macro 'INTERCEPT_FIELD'
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/501/e400fe7283b021af3e4996796e51346e481c204c/fullLog.txt).

--- a/Utilities/MCStepLogger/src/MCStepInterceptor.cxx
+++ b/Utilities/MCStepLogger/src/MCStepInterceptor.cxx
@@ -26,44 +26,78 @@
 //  @brief  A LD_PRELOAD logger hooking into Stepping of TVirtualMCApplication's
 
 class TVirtualMCApplication;
+class TVirtualMagField;
 
 // (re)declare symbols to be able to hook into them
-#define DECLARE_INTERCEPT_SYMBOLS(APP)                                                       \
-  class APP                                                                                  \
-  {                                                                                          \
-     public:                                                                                 \
-       void Stepping();                                                                      \
-       void FinishEvent();                                                                   \
+#define DECLARE_INTERCEPT_SYMBOLS(APP) \
+  class APP                            \
+  {                                    \
+   public:                             \
+    void Stepping();                   \
+    void FinishEvent();                \
   };
 
 DECLARE_INTERCEPT_SYMBOLS(FairMCApplication)
 DECLARE_INTERCEPT_SYMBOLS(AliMC)
 
+// same for field
+#define DECLARE_INTERCEPT_FIELD_SYMBOLS(FIELD)                 \
+  class FIELD                                                  \
+  {                                                            \
+   public:                                                     \
+    void GetFieldValue(const double point[3], double* bField); \
+  };
+
+namespace o2
+{
+namespace field
+{
+DECLARE_INTERCEPT_FIELD_SYMBOLS(MagneticField);
+}
+}
 
 extern "C" void performLogging(TVirtualMCApplication*);
-extern "C" void dispatchOriginal(TVirtualMCApplication*, char const *libname, char const*);
+extern "C" void logField();
+extern "C" void dispatchOriginal(TVirtualMCApplication*, char const* libname, char const*);
+extern "C" void dispatchOriginalField(TVirtualMagField*, char const* libname, char const*, const double x[3],
+                                      double* B);
 extern "C" void flushLog();
 
-#define INTERCEPT_STEPPING(APP, LIB, SYMBOL)                                                 \
-  void APP::Stepping()                                                                       \
-  {                                                                                          \
-     auto baseptr=reinterpret_cast<TVirtualMCApplication*>(this);                            \
-     performLogging(baseptr);                                                                \
-     dispatchOriginal(baseptr, LIB, SYMBOL);                                                 \
-  }                                                                                          \
-
-#define INTERCEPT_FINISHEVENT(APP, LIB, SYMBOL)                                              \
-  void APP::FinishEvent()                                                                    \
-  {                                                                                          \
-     auto baseptr=reinterpret_cast<TVirtualMCApplication*>(this);                            \
-     flushLog();                                                                             \
-     dispatchOriginal(baseptr, LIB, SYMBOL);                                                 \
+#define INTERCEPT_STEPPING(APP, LIB, SYMBOL)                       \
+  void APP::Stepping()                                             \
+  {                                                                \
+    auto baseptr = reinterpret_cast<TVirtualMCApplication*>(this); \
+    performLogging(baseptr);                                       \
+    dispatchOriginal(baseptr, LIB, SYMBOL);                        \
   }
 
+#define INTERCEPT_FINISHEVENT(APP, LIB, SYMBOL)                    \
+  void APP::FinishEvent()                                          \
+  {                                                                \
+    auto baseptr = reinterpret_cast<TVirtualMCApplication*>(this); \
+    flushLog();                                                    \
+    dispatchOriginal(baseptr, LIB, SYMBOL);                        \
+  }
 
 // the runtime will now dispatch to these functions due to LD_PRELOAD
-INTERCEPT_STEPPING(FairMCApplication,"libBase","_ZN17FairMCApplication8SteppingEv")
-INTERCEPT_STEPPING(AliMC,"libSTEER","_ZN5AliMC8SteppingEv")
+INTERCEPT_STEPPING(FairMCApplication, "libBase", "_ZN17FairMCApplication8SteppingEv")
+INTERCEPT_STEPPING(AliMC, "libSTEER", "_ZN5AliMC8SteppingEv")
 
-INTERCEPT_FINISHEVENT(FairMCApplication,"libBase","_ZN17FairMCApplication11FinishEventEv")
-INTERCEPT_FINISHEVENT(AliMC,"libSTEER","_ZN5AliMC11FinishEventEv")
+INTERCEPT_FINISHEVENT(FairMCApplication, "libBase", "_ZN17FairMCApplication11FinishEventEv")
+INTERCEPT_FINISHEVENT(AliMC, "libSTEER", "_ZN5AliMC11FinishEventEv")
+
+#define INTERCEPT_FIELD(FIELD, LIB, SYMBOL)                        \
+  void FIELD::GetFieldValue(const double point[3], double* bField) \
+  {                                                                \
+    auto baseptr = reinterpret_cast<TVirtualMagField*>(this);      \
+    logField();                                                    \
+    dispatchOriginalField(baseptr, LIB, SYMBOL, point, bField);    \
+  }
+
+namespace o2
+{
+namespace field
+{
+INTERCEPT_FIELD(MagneticField, "libField", "_ZN2o25field13MagneticField13GetFieldValueEPKdPd");
+}
+}

--- a/Utilities/MCStepLogger/src/MCStepLoggerImpl.cxx
+++ b/Utilities/MCStepLogger/src/MCStepLoggerImpl.cxx
@@ -25,10 +25,11 @@
 //  @since  2017-06-29
 //  @brief  A logging service for MCSteps (hooking into Stepping of TVirtualMCApplication's)
 
-#include <TVirtualMCApplication.h>
-#include <TVirtualMC.h>
-#include <sstream>
 #include <TTree.h>
+#include <TVirtualMC.h>
+#include <TVirtualMCApplication.h>
+#include <TVirtualMagField.h>
+#include <sstream>
 
 #include <dlfcn.h>
 #include <iostream>
@@ -39,7 +40,50 @@
 #endif
 #include <cassert>
 
-namespace o2 {
+namespace o2
+{
+// a class collecting field access per volume
+class FieldLogger
+{
+  int counter = 0;
+  std::map<int, int> volumetosteps;
+  std::map<int, std::string> idtovolname;
+
+ public:
+  void addStep(TVirtualMC* mc)
+  {
+    counter++;
+    int copyNo;
+    auto id = mc->CurrentVolID(copyNo);
+    if (volumetosteps.find(id) == volumetosteps.end()) {
+      volumetosteps.insert(std::pair<int, int>(id, 0));
+    } else {
+      volumetosteps[id]++;
+    }
+    if (idtovolname.find(id) == idtovolname.end()) {
+      idtovolname.insert(std::pair<int, std::string>(id, std::string(mc->CurrentVolName())));
+    }
+  }
+
+  void clear()
+  {
+    counter = 0;
+    volumetosteps.clear();
+    idtovolname.clear();
+  }
+
+  void flush()
+  {
+    std::cerr << "[FIELDLOGGER]: did " << counter << " steps \n";
+    // summarize steps per volume
+    for (auto& p : volumetosteps) {
+      std::cerr << "[FIELDLOGGER]: VolName " << idtovolname[p.first] << " COUNT " << p.second;
+      std::cerr << "\n";
+    }
+    clear();
+    std::cerr << "[FIELDLOGGER]: ----- END OF EVENT ------\n";
+  }
+};
 
 class StepLogger
 {
@@ -48,7 +92,9 @@ class StepLogger
   std::set<int> trackset;
   std::set<int> pdgset;
   std::map<int, int> volumetosteps;
-  std::map<int, char const*> idtovolname;
+  std::map<int, std::string> idtovolname;
+  std::map<int, int> volumetoNSecondaries;            // number of secondaries created in this volume
+  std::map<std::pair<int, int>, int> volumetoProcess; // mapping of volumeid x processID to secondaries produced
 
   // TODO: consider writing to a TTree/TFile
  public:
@@ -56,88 +102,143 @@ class StepLogger
   {
     assert(mc);
     stepcounter++;
-    auto stack=mc->GetStack();
+    auto stack = mc->GetStack();
     assert(stack);
     trackset.insert(stack->GetCurrentTrackNumber());
     pdgset.insert(mc->TrackPid());
     int copyNo;
     auto id = mc->CurrentVolID(copyNo);
-    if (volumetosteps.find(id) == volumetosteps.end()){
+    if (volumetosteps.find(id) == volumetosteps.end()) {
       volumetosteps.insert(std::pair<int, int>(id, 0));
-    }
-    else {
+    } else {
       volumetosteps[id]++;
     }
     if (idtovolname.find(id) == idtovolname.end()) {
-      idtovolname.insert(std::pair<int, char const*>(id, mc->CurrentVolName()));
+      idtovolname.insert(std::pair<int, std::string>(id, std::string(mc->CurrentVolName())));
+    }
+
+    // for the secondaries
+    if (volumetoNSecondaries.find(id) == volumetoNSecondaries.end()) {
+      volumetoNSecondaries.insert(std::pair<int, int>(id, mc->NSecondaries()));
+    } else {
+      volumetoNSecondaries[id] += mc->NSecondaries();
+    }
+
+    // for the processes
+    for (int i = 0; i < mc->NSecondaries(); ++i) {
+      auto process = mc->ProdProcess(i);
+      auto p = std::pair<int, int>(id, process);
+      if (volumetoProcess.find(p) == volumetoProcess.end()) {
+        volumetoProcess.insert(std::pair<std::pair<int, int>, int>(p, 1));
+      } else {
+        volumetoProcess[p]++;
+      }
     }
   }
 
-  void clear() {
+  void clear()
+  {
     stepcounter = 0;
     trackset.clear();
     pdgset.clear();
     volumetosteps.clear();
     idtovolname.clear();
+    volumetoNSecondaries.clear();
+    volumetoProcess.clear();
   }
 
-  void flush() {
-    std::cerr << "did " << stepcounter << " steps \n";
-    std::cerr << "transported " << trackset.size() << " different tracks \n";
-    std::cerr << "transported " << pdgset.size() << " different types \n";
+  // prints list of processes for volumeID
+  void printProcesses(int volid)
+  {
+    for (auto& p : volumetoProcess) {
+      if (p.first.first == volid) {
+        std::cerr << "P[" << TMCProcessName[p.first.second] << "]:" << p.second << "\t";
+      }
+    }
+  }
+
+  void flush()
+  {
+    std::cerr << "[STEPLOGGER]: did " << stepcounter << " steps \n";
+    std::cerr << "[STEPLOGGER]: transported " << trackset.size() << " different tracks \n";
+    std::cerr << "[STEPLOGGER]: transported " << pdgset.size() << " different types \n";
     // summarize steps per volume
     for (auto& p : volumetosteps) {
-      std::cout << " VolName " << idtovolname[p.first] << " COUNT " << p.second << "\n";
+      std::cerr << "[STEPLOGGER]: VolName " << idtovolname[p.first] << " COUNT " << p.second << " SECONDARIES "
+                << volumetoNSecondaries[p.first] << " ";
+      // loop over processes
+      printProcesses(p.first);
+      std::cerr << "\n";
     }
     clear();
+    std::cerr << "[STEPLOGGER]: ----- END OF EVENT ------\n";
   }
 };
 
 StepLogger logger;
+FieldLogger fieldlogger;
 
 } // end namespace
 
-// a generic function that can dispatch to the original method of a TVirtualMCApplication
-// (for functions of type void TVirtualMCApplication::Method(void); )
-extern "C" void dispatchOriginal(TVirtualMCApplication* app, char const* libname, char const* origFunctionName)
+// a helper template kernel describing generically the redispatching prodecure
+template <typename Object /* the original object type */, typename MethodType /* member function type */,
+          typename... Args /* original arguments to function */>
+void dispatchOriginalKernel(Object* obj, char const* libname, char const* origFunctionName, Args... args)
 {
-  typedef void (TVirtualMCApplication::*StepMethodType)();
+  // Object, MethodType, and Args are of course related so we could do some static_assert checks or automatic deduction
+
   // static map to avoid having to lookup the right symbols in the shared lib at each call
   // (We could do this outside of course)
-  static std::map<const char *, StepMethodType> functionNameToSymbolMap;
-  StepMethodType origMethod = nullptr;
+  static std::map<const char*, MethodType> functionNameToSymbolMap;
+  MethodType origMethod = nullptr;
 
   auto iter = functionNameToSymbolMap.find(origFunctionName);
-  if(iter == functionNameToSymbolMap.end()){
+  if (iter == functionNameToSymbolMap.end()) {
     auto libHandle = dlopen(libname, RTLD_NOW);
     // try to make the library loading a bit more portable:
-    if (!libHandle){
+    if (!libHandle) {
       // try appending *.so
       std::stringstream stream;
       stream << libname << ".so";
       libHandle = dlopen(stream.str().c_str(), RTLD_NOW);
-    } 
-    if (!libHandle){
+    }
+    if (!libHandle) {
       // try appending *.dylib
       std::stringstream stream;
       stream << libname << ".dylib";
       libHandle = dlopen(stream.str().c_str(), RTLD_NOW);
-    } 
+    }
     assert(libHandle);
     void* symbolAddress = dlsym(libHandle, origFunctionName);
     assert(symbolAddress);
-    // Purposely ignore compiler warning
+// Purposely ignore compiler warning
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wsizeof-pointer-memaccess"
     // hack since C++ does not allow casting to C++ member function pointers
     // thanks to gist.github.com/mooware/1174572
     memcpy(&origMethod, &symbolAddress, sizeof(&symbolAddress));
 #pragma GCC diagnostic pop
-    functionNameToSymbolMap[origFunctionName]=origMethod;
+    functionNameToSymbolMap[origFunctionName] = origMethod;
   } else {
     origMethod = iter->second;
   }
-  (app->*origMethod)();
+  // the final C++ member function call redispatch
+  (obj->*origMethod)(args...);
+}
+
+// a generic function that can dispatch to the original method of a TVirtualMCApplication
+extern "C" void dispatchOriginal(TVirtualMCApplication* app, char const* libname, char const* origFunctionName)
+{
+  typedef void (TVirtualMCApplication::*StepMethodType)();
+  dispatchOriginalKernel<TVirtualMCApplication, StepMethodType>(app, libname, origFunctionName);
+}
+
+// a generic function that can dispatch to the original method of a TVirtualMagField
+extern "C" void dispatchOriginalField(TVirtualMagField* field, char const* libname, char const* origFunctionName,
+                                      const double x[3], double* B)
+{
+  typedef void (TVirtualMagField::*MethodType)(const double[3], double*);
+  dispatchOriginalKernel<TVirtualMagField, MethodType>(field, libname, origFunctionName, x, B);
 }
 
 extern "C" void performLogging(TVirtualMCApplication* app)
@@ -146,6 +247,14 @@ extern "C" void performLogging(TVirtualMCApplication* app)
   o2::logger.addStep(mc);
 }
 
-extern "C" void flushLog(){
+extern "C" void logField()
+{
+  static TVirtualMC* mc = TVirtualMC::GetMC();
+  o2::fieldlogger.addStep(mc);
+}
+
+extern "C" void flushLog()
+{
   o2::logger.flush();
+  o2::fieldlogger.flush();
 }

